# Within-host genetic diversity of SARS-CoV-2 lineages in unvaccinated and vaccinated individuals

*Haogao Gu, Ahmed Abdul Quadeer, Pavithra Krishnan, Daisy Y.M. Ng, Lydia D.J Chang, Gigi Y.Z. Liu, Samuel S.M. Cheng, Tommy T.Y. Lam, Malik Peiris, Matthew R. McKay, Leo L.M. Poon**

### Abstract 

Viral and host factors can shape SARS-CoV-2 evolution. However, little is known about lineage-specific and vaccination-specific mutations that occur within individuals. Here, we analysed deep sequencing data from 2,820 SARS-CoV-2 respiratory samples with different viral lineages to describe the patterns of within-host diversity under different conditions, including vaccine-breakthrough infections. In unvaccinated individuals, variant of Concern (VOC) Alpha, Delta, and Omicron respiratory samples were found to have higher within-host diversity and were under neutral to purifying selection at the full genome level compared to non-VOC SARS-CoV-2. Breakthrough infections in 2-dose or 3-dose Comirnaty and CoronaVac vaccinated individuals did not increase levels of non-synonymous mutations and did not change the direction of selection pressure. Vaccine-induced antibody or T cell responses did not appear to have significant impact on within-host SARS-CoV-2 sequence diversification. Our findings suggest that vaccination does not increase exploration of SARS-CoV-2 protein sequence space and may not facilitate emergence of viral variants.

### Data availability
The sequencing data used in this study can be access through NCBI Sequence Read Archive (SRA) with Bioproject accession ID: PRJNA930974. The anonymised metadata are deposited at https://github.com/Leo-Poon-Lab/mutations-under-sarscov2-vaccination/tree/main/metadata.

### Code availability
Detailed scripts for the above analysis are available from https://github.com/Leo-Poon-Lab/mutations-under-sarscov2-vaccination.



